### PR TITLE
fix enum down casting and backtrace panics

### PR DESCRIPTION
**Pull request by oli-obk**
enum downcasting assumed that the actual variant data comes directly after the discriminant, but that's obviously wrong in presence of alignment and thus padding.

### CODE REVIEW

**Comment by oli-obk** [comment]
The following doesn't work yet

```rust
#![feature(lang_items, unboxed_closures, fn_traits)]

struct S3 {
    x: i32,
    y: i32,
}

impl FnOnce<(i32,i32)> for S3 {
    type Output = i32;
    extern "rust-call" fn call_once(self, (z,zz): (i32,i32)) -> i32 {
        self.x * self.y * z * zz
    }
}

fn main() {
    let s = S3 {
        x: 3,
        y: 3,
    };
    let ans = s(3, 1);
    assert_eq!(ans, 27);
}
```

it fails, because the call splits the arguments into single arguments and there's no shim to re-tuple them. We should probably just skip the splitting when the actual function we are calling isn't a closure, but a custom `impl Fn*...`.

File: tests/run-pass/try-operator-custom.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+    assert!(Ok::<i32, String>(42) == Ok(42));
```
Comment (solson) [inline tests/run-pass/try-operator-custom.rs]: How does this test the backtrace panic? Shouldn't we have a compile-fail test that checks for the proper backtrace notes?

Comment (oli-obk) [inline tests/run-pass/try-operator-custom.rs]: No this test should succeed and does succeed. I can add some tests that fail, but to reach large backtraces we probably need UB

File: src/interpreter/terminator/mod.rs [hunk 2]
```diff
@@ -388,8 +388,41 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
+                        //
+                        // These are both the same at trans time.
+
+                        // interpreter magic: insert an intermediate pointer, so we can skip the intermediate function call
+                        // FIXME: this is a memory leak, should probably add the pointer to the current stack
+                        let ptr_size = self.memory.pointer_size();
+                        let first = self.memory.allocate(ptr_size, ptr_size)?;
+                        self.memory.copy(args[0].0, first, ptr_size, ptr_size)?;
+                        self.memory.write_ptr(args[0].0, first)?;
+                        self.memory.dump(args[0].0.alloc_id);
```
Comment (solson) [inline src/interpreter/terminator/mod.rs]: Leftover debug print.

**Pull request merged**